packages/@react-stately/combobox/src/useComboBoxState.ts

```diff
@@ -0,0 +1,304 @@
+  /*
+  let menuControlled = props.isOpen !== undefined;
+  let valueControlled = inputValue !== undefined;
+  let selectedControlled = !!selectedKey;
+   */
+
+  // Create a separate menu open state tracker so onOpenChange doesn't fire with open and close in quick succession
+  // in cases where there aren't items to show
+  // Note that this means onOpenChange won't fire for controlled open states
+  let [menuIsOpen, setMenuIsOpen] = useControlledState(isOpen, defaultOpen || false, () => {});
+
+  let selectState = useSelectState(props);
+
+  let selectedKeyItem = selectedKey ? selectState.collection.getItem(selectedKey) : undefined;
+  let selectedKeyText = selectedKeyItem ? selectedKeyItem.textValue || selectedKeyItem.rendered as string : undefined;
+  // Maybe don't need to do for defaultSelectedKey? Need to do for selectedKey so that the textfield is properly controlled reflects selectedKey text
+  let defaultSelectedKeyItem = defaultSelectedKey ? selectState.collection.getItem(defaultSelectedKey) : undefined;
+  let defaultSelectedKeyText = defaultSelectedKeyItem ? defaultSelectedKeyItem.textValue || defaultSelectedKeyItem.rendered as string : undefined;
+  // Double check if selectedKey should make textfield value controlled
+  let [value, setValue] = useControlledState(toString(inputValue) || selectedKeyText, toString(defaultInputValue) || defaultSelectedKeyText || '', onInputChange);
+  let lowercaseValue = value.toLowerCase();
+
+  let defaultFilterFn = useMemo(() => (node: Node<T>) => {
+    let scan = 0;
+    let lowercaseNode = node.textValue.toLowerCase();
+    for (let i in lowercaseValue) {
+      if (whitespace.test(lowercaseValue[i])) {
+        continue;
+      }
+      let match = false;
+      for (; scan < lowercaseNode.length && !match; scan++) {
+        if (!whitespace.test(lowercaseNode[scan])) { // if whitespace, move on to next character
+          let compareVal = collator.compare(lowercaseValue[i], lowercaseNode[scan]);
+          if (compareVal !== 0) { // if it didn't match, we can exit
+            return false;
+          } else { // we matched, move to next character in input for comparison
+            match = true;
+          }
+        }
+      }
+    }
+    return true;
+  }, [collator, lowercaseValue]);
+
+  useEffect(() => {
+    if (onFilter) {
+      onFilter(value);
+    }
+  // Having onFilter in the dep array seems to break it
+  // eslint-disable-next-line react-hooks/exhaustive-deps
+  }, [value]);
+
+  selectState.collection = useMemo(() => {
+    if (itemsControlled || value === '') {
+      return selectState.collection;
+    }
+    return new FilteredCollection(selectState.collection, defaultFilterFn);
+  }, [selectState.collection, value, itemsControlled, defaultFilterFn]);
+ 
+  useEffect(() => {
+    // Close the menu if it was open but there aren't any items to display
+    if (menuIsOpen && selectState.collection.size === 0) {
+      selectState.close();
```

Comment:
What do you think of not updating state at all for these cases, and instead computing whether state is open on render. e.g. when returning `isOpen` below, you could do something like:

```js
isOpen: selectState.isOpen && isFocused && selectState.collection.size > 0
```

This way you wouldn't need to update state in more than one place and could probably get rid of this `useEffect`.

This might also allow you to get rid of the extra copy of state you have `menuIsOpen` and keep things centralized in one place.

Reply:
oo, fantastic idea! I'll give it a go